rust/arrow/src/buffer.rs

```diff
@@ -756,15 +738,15 @@ impl MutableBuffer {
     /// also ensure the new capacity will be a multiple of 64 bytes.
     ///
     /// Returns the new capacity for this buffer.
-    pub fn reserve(&mut self, capacity: usize) -> usize {
-        if capacity > self.capacity {
-            let new_capacity = bit_util::round_upto_multiple_of_64(capacity);
-            let new_capacity = cmp::max(new_capacity, self.capacity * 2);
+    pub fn reserve(&mut self, additional: usize) {
```

Comment:
Is there any reason to change the API in this PR? don't see how it could be related to the performance issue.
Also better to split it to make PR easier to review. We also need to update the doc.

Reply:
I agree that it makes it more difficult. The signature is related to performance, though:

It was difficult to reason with `reserve(new_len)` when `Vec::reserve` and `RawVec::reserve` uses `additional` (and I was basing this PR on that). This PR fixes a bug in one `Builder` that was calling `reserve(1)` in `append`, when it should be calling `reserve(self.len + 1)`. I also spent 4hs tracking a bug because of this difference (I was using `reserve(additional)` when its signature was `reserve(new_len)`.

I can try to split that to another PR, though. IMO this PR will still need to be based on that PR.

Reply:
I think originally we referred [C++ impl](https://github.com/apache/arrow/blob/master/cpp/src/arrow/buffer.h#L436) for the signature design to make it consistent. Since `buffer` module is public, we'll need to be careful when making any breaking change on API.

Reply:
Yap, and I think it was a good decision back then, when the goal was to port C++ code to Rust. I understand your concern. 

However, IMO we should not keep using C++-motivated APIs because it was historically like that. `additional` is the standard in virtually every container in Rust (`HashMap`, `Vec`, `tokio::bytes::Bytes`). IMO this change will need to be made at some point, and delaying it will only cause more pain in the future.

In this particular instance, it was motivated by an already existing confusion on our own code base, which IMO evidences that rust developers expect `additional`. Most people that use this API are probably passing `additional` to it already.

Regardless, we are bumping the major version, so there is already no expectation of backward compatibility as far as cargo and semver is concerned. My goal here is to make `MutableBuffer` as close as possible to `Vec` so that people can use them without even thinking that they are using a special allocation.

Reply:
Thanks, if we're going to bump the major version then I'm fine with the compatibility issue. I don't have strong opinion on this so either is fine to me. IMO a good documentation should be suffice for developers to understand what the method does.